### PR TITLE
Duoswap Interactive

### PR DESCRIPTION
**Pull request by chrisnevers**
Interactive command line version of Duoswap. Ignoring this example in CI since it is interactive and there's an automated version already.

How I run it:

1. Run Announcer Manager
  * Creates contract that listens for pool information and stores info locally in a cache (javascript object)
2. Run Funder
  * Creates 2 tokens that will be used in the duoswap pool.
  * Enter the address of an account via CLI to fund it
3. Run DuoSwap Admin
  * Creates pool & sends message to announcer contract with pool address
4. Run DuoSwap Liquidity Provider
  * Make initial deposit into pool
5. Run DuoSwap Trader
  * Makes a trade
 
The one thing less than ideal about running this is whenever someone misses a race, i.e. another participant performs an action, the current CLI prompt is for the missed race. So if a Trader is asked to trade, and a LP makes a deposit in the mean time, the response to the trader's prompt is useless and has to be re-entered the _next_ time they are asked.

### CODE REVIEW

File: .circleci/xfail.txt [hunk 1]
```diff
@@ -9,6 +9,11 @@ ETH.tut-7-rpc
 ALGO.tut-7-rpc
 CFX.tut-7-rpc
 
+# Interactive
```
Comment (jeapostrophe) [inline .circleci/xfail.txt:12]: You can look at how `tut-8` works and I script an interactive program for testing.

Comment (jeapostrophe) [reply .circleci/xfail.txt:12]: Alternatively, you could just make an interactive mode for `examples/duoswap/index.mjs`

File: examples/duoswap-interactive/index.mjs [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+  '1: DuoSwap Admin',
+  '2: DuoSwap Liquidity Provider',
+  '3: DuoSwap Trader',
+  '4: DuoSwap Announcer',
+  '5: DuoSwap Tokens',
+].join('\n');
+
+const answer = await ask.ask(`Who are you?\n${options}`, parseInt);
+
+if (answer == 1) {
```
Comment (jeapostrophe) [inline examples/duoswap-interactive/index.mjs:188]: You should do a `switch`

File: examples/duoswap-interactive/index.rsh [hunk 3]
```diff
@@ -0,0 +1,296 @@
+'reach 0.1';
```
Comment (jeapostrophe) [inline examples/duoswap-interactive/index.rsh:1]: I don't like that this file is duplicated. I think that you should just make an interactive version in the other directory

File: examples/duoswap-interactive/announcer.rsh [hunk 4]
```diff
@@ -0,0 +1,49 @@
+'reach 0.1';
```
Comment (jeapostrophe) [inline examples/duoswap-interactive/announcer.rsh:1]: Will the `hs-test` look at this? Doesn't it need an `announcer.txt` file?

File: .circleci/xfail.txt (hunk 1 above)
Comment (jeapostrophe) [reply .circleci/xfail.txt:12]: At the very least, you should have a README explaining how to test interactively and what to expect to see

**Review by jeapostrophe — CHANGES_REQUESTED** [review]

File: .circleci/xfail.txt (hunk 1 above)
Comment (chrisnevers) [reply .circleci/xfail.txt]: Made an option to run interactively in the original `duoswap` directly. The test will just run the automated version. Did a Readme, too

File: examples/duoswap/index.rsh [hunk 5]
```diff
@@ -112,6 +123,8 @@ export const main = Reach.App(() => {
     parallelReduce([ true, initialMarket, 0 ])
       .define(() => {
         const st = [ alive, market ];
+        // Tokens.aBal.set(balance(tokA));
```
Comment (chrisnevers) [inline examples/duoswap/index.rsh:126]: The frontend isn't using this info right now and commenting out these views shave 2 minutes off of compilation time

File: .circleci/xfail.txt (hunk 1 above)
Comment (jeapostrophe) [reply .circleci/xfail.txt]: Can the README have a sample interactive session log?

**Comment by jonmoshier** [comment]
I get an error when running duoswap from a fresh repo.

### Steps to reproduce:
1. `git clone` the reach-lang repository.
2. cd into `\reach-lang\examples\duoswap`
3. curl and chmod `reach` binary
4. `./reach run -- i`

reach successfully compiles `index.rsh`, and ONLY `index.rsh`, and runs but then hits an error

```
internal/process/esm_loader.js:74
    internalBinding('errors').triggerUncaughtException(
                              ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/app/build/announcer.main.mjs' imported from /app/announcer.mjs
```

The issue here is that announcer.rsh hasn't been built.

### Suggested fixes:
1. Either add a step to the README that tells the user to run `reach compile` against announcer and other reach files 

OR

2. Include compilation as part of `reach run`

**Comment by jeapostrophe** [comment]
Run `make` first

**Pull request merged**